src/compiler/lowering/MatMulOpLowering.cpp

```diff
@@ -14,198 +14,564 @@
+                       mlir::MLIRContext *ctx,
+                       llvm::SmallVector<AffineForOp, 3> &loops,
+                       Type elementType, int64_t vec_size) {
+  auto vec_Type = mlir::VectorType::get({vec_size}, elementType);
+
+  // TODO: We need an option to enable smaller vector sizes for the ends of each
+  // row. row loop
+  auto rowLoop = rewriter.create<AffineForOp>(loc, 0, lhsShape[ROW], 1);
+  // row loop body
+  rewriter.setInsertionPointToStart(rowLoop.getBody());
+  // col loop
+  auto colLoop = rewriter.create<AffineForOp>(loc, 0, rhsShape[COL], vec_size);
+  // col loop body
+  rewriter.setInsertionPointToStart(colLoop.getBody());
+  // fma loop
+  auto fmaLoop = rewriter.create<AffineForOp>(loc, 0, rhsShape[ROW], 1);
+  // inner loop body
+  rewriter.setInsertionPointToStart(fmaLoop.getBody());
+
+  auto a_single = rewriter.create<AffineLoadOp>(
+      loc, lhs,
+      ValueRange{rowLoop.getInductionVar(), fmaLoop.getInductionVar()});
+  auto a = rewriter.create<vector::SplatOp>(loc, a_single, vec_Type);
+  auto b = rewriter.create<AffineVectorLoadOp>(
+      loc, vec_Type, rhs,
+      ValueRange{fmaLoop.getInductionVar(), colLoop.getInductionVar()});
+  auto c = rewriter.create<AffineVectorLoadOp>(
+      loc, vec_Type, output,
+      ValueRange{rowLoop.getInductionVar(), colLoop.getInductionVar()});
+
+  Value res = rewriter.create<vector::FMAOp>(loc, a, b, c);
+
+  rewriter.create<AffineVectorStoreOp>(
+      loc, res, output,
+      ValueRange{rowLoop.getInductionVar(), colLoop.getInductionVar()});
+
+  // AffineYieldOp at end of loop blocks
+  rewriter.setInsertionPointAfter(fmaLoop);
+  rewriter.setInsertionPointAfter(colLoop);
+  rewriter.setInsertionPointAfter(rowLoop);
+
+  loops.push_back(rowLoop);
+  loops.push_back(colLoop);
+  loops.push_back(fmaLoop);
+  return loops;
+}
 
-    // load
-    mlir::Value a = rewriter.create<memref::LoadOp>(
-        loc, lhs, ValueRange{loopIvs[0], loopIvs[2]});
-    mlir::Value b = rewriter.create<memref::LoadOp>(
-        loc, rhs, ValueRange{loopIvs[2], loopIvs[1]});
-    mlir::Value c = rewriter.create<memref::LoadOp>(
-        loc, output, ValueRange{loopIvs[0], loopIvs[1]});
+class MatMulLowering : public OpConversionPattern<daphne::MatMulOp> {
+  const LowerMatMulOpOptions options;
+
+public:
+  using OpConversionPattern::OpConversionPattern;
+  explicit MatMulLowering(MLIRContext *context,
+                          LowerMatMulOpOptions const &options)
+      : OpConversionPattern(context, PatternBenefit(1)), options(options) {}
+
+  bool is_vectorizable(ArrayRef<int64_t> const rhsShape,
+                       Type const matrixElementType) const {
+    if (rhsShape[COL] %
+            options.getVecSize(matrixElementType.getIntOrFloatBitWidth()) !=
+        0) {
+      return false;
+    }
+    if (!matrixElementType.isa<FloatType>()) {
+      return false;
+    }
+    return true;
+  }
+
+  // TODO: Test why not
+  // Cannot tile non square matmul.
```

Comment:
This would be nice to have, especially for your benchmarks. In case you run into too much troubles fixing this, handle it in the same way as the `TODO` above.

Reply:
Seems to work currently also for non square matrices. I will do some more testing.